### DISCUSSION TITLE
Code targeting multiple runtimes

### DISCUSSION THREAD

**Discussion by Gozala** [jaredly/jerd#3] [General]
Hey all,

Lately I've been dealing with lot of JS code that needs to use different implementation of a function based on the target runtime (nodejs, web, electron, etc...). Usually in these scenarios you'd have:

1. Same interface / type for it.
2. Different implementation per target runtime.
3. Some static way to specify which implementation is for which runtime.
4. Compiler / bundler / linker that producing result with runtime based substitutions.

I think it would be a really useful to have a good solution in this language, furthermore given a type system it could probably do far better than user space solutions prominent in JS.

When I think about it seems like either:

1. A term references not by a single hash but by a set of hashes. Selecting a concrete hash at compiler time based on target.
2. Something like terms generic over  runtime. Although not sure how that could manifest.

I think it is also worth considering that in some instances runtime based substitution is needed, e.g. when you need to check if a specific feature is available in the web runtime you're running on. I think it would be really interesting to have well thought out solution for this as well.

Posting it here to query for thoughts

**Comment by jaredly** [comment]
Could you come up with some specific examples?
My assumption is that the places where you would want to have different behavior based on platform are all effectful ones (interacting with the outside world), and so algebraic effects give us this split for free! (e.g. the "host" platform implements handlers for the relevant effects, and so the code doesn't have to care about platform differences).
